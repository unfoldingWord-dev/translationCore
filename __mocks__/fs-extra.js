'use strict';
import path from 'path-extra';
import _ from "lodash";
const fsActual = require.requireActual('fs-extra'); // for copying test files into mock
const fs = jest.genMockFromModule('fs-extra');
let mockFS = Object.create(null);

/** @deprecated */
function __setMockFS(newMockFS) {
  mockFS = newMockFS;
}

function __resetMockFS() {
  mockFS = Object.create(null);
}

/**
 * This is a custom function that our tests can use during setup to specify
 * what the files on the "mock" filesystem should look like when any of the
 * `fs` APIs are used.
 * @param {Array} newMockFiles
 */
function __setMockDirectories(newMockFiles) {
  mockFS = Object.create(null);
  for (const file in newMockFiles) {
    const dir = path.dirname(file);

    if (!mockFS[dir]) {
      mockFS[dir] = [];
    }
    mockFS[dir].push(path.basename(file));
  }
}

/**
 * A custom version of `readdirSync` that reads from the special mocked out
 * file list set via __setMockDirectories
 * @param {String} directoryPath
 */
function readdirSync(directoryPath) {
  if (statSync(directoryPath).isDirectory()) {
    return mockFS[directoryPath];
  }
  return [];
}

function readdir(directoryPath) {
  return Promise.resolve(readdirSync(directoryPath));
}

function writeFileSync(filePath, data) {
  addFileToParentDirectory(filePath);
  mockFS[filePath] = data;
}

function readFileSync(filePath) {
  if (typeof filePath !== 'string') throw 'fail';
  const data = mockFS[filePath];
  // TRICKY: readFileSync should always return a string
  if(typeof data === 'object' && data !== null) {
    return JSON.stringify(data);
  } else {
    return data;
  }
}

function outputFileSync(filePath, data) {
  addFileToParentDirectory(filePath);
  mockFS[filePath] = data;
}

function __dumpMockFS() {
  const fsList = JSON.stringify(mockFS, null, 2);
  console.log("mock FS:\n" + fsList);
}

/**
 * create subdirs and add file name to them
 * @param filePath
 */
function addFileToParentDirectory(filePath) {
  const dir = path.dirname(filePath);
  const filename = path.basename(filePath);
  if (filename) {
    if (!mockFS[dir]) {
      mockFS[dir] = [];
      addFileToParentDirectory(dir);
    }
    if (mockFS[dir].indexOf(filename) < 0) {
      mockFS[dir].push(filename);
    }
  }
}

function outputJsonSync(filePath, data) {
  addFileToParentDirectory(filePath);
  mockFS[filePath] = _.cloneDeep(data);
}

function outputJson(filePath, data) {
  return new Promise(function (resolve) {
    outputJsonSync(filePath, data);
    resolve();
  });
}

function readJsonSync(filePath) {
  if (!existsSync(filePath)) {
    throw 'File could not be read: ' + filePath;
  }

  const data = mockFS[filePath];
  // clone data so changes to object do not affect object in file system
  const clonedData = JSON.parse(typeof data === 'string' ? data : JSON.stringify(data));
  return clonedData;
}

function readJson(filePath) {
  return Promise.resolve(readJsonSync(filePath));
}

function existsSync(path) {
  const exists = mockFS[path] !== '' ? !!mockFS[path] : true;
  return exists;
}

function exists(path) {
  return Promise.resolve(existsSync(path));
}

function removeSync(path) {
  Object.keys(mockFS).forEach((element) => {
    element.includes(path) ? delete mockFS[element] : null;
  });
}

function remove(path) {
  return new Promise(function (resolve) {
    removeSync(path);
    resolve();
  });
}

function renameSync(oldPath, newPath) {
  writeFileSync(newPath, readFileSync(oldPath));
  removeSync(oldPath);
}

function rename(oldPath, newPath) {
  return new Promise(function (resolve) {
    renameSync(oldPath, newPath);
    resolve();
  });
}

function copySync(srcPath, destinationPath) {
  mockFS[destinationPath] = mockFS[srcPath];
  addFileToParentDirectory(destinationPath);
  const isDir = statSync(srcPath).isDirectory();
  if (isDir) {
    const files = readdirSync(srcPath);
    for (let f of files) {
      copySync(path.join(srcPath,f), path.join(destinationPath,f));
    }
  }
}

function ensureDirSync(path) {
  if (!mockFS[path]) mockFS[path] = [];
  addFileToParentDirectory(path);
}

function Stats(path, exists, isDir) {
  this.path = path;
  this.exists = exists;
  this.isDir = isDir;
  this.atime = "Not-a-real-date";
  this.isDirectory = () => {
    const isDir = this.exists && this.isDir;
    return isDir;
  };
  this.isFile = () => {
    const isFile = this.exists && !this.isDir;
    return isFile;
  };
}

/**
 * ensure this actually contains an array of file names (strings)
 * @param path
 * @return {arg is Array<any>}
 */
function isValidDirectory(path) {
  const dir = mockFS[path];
  let isDir = Array.isArray(dir);
  if (isDir) { // make sure it's an array of paths (strings) and not objects (such as json object stored)
    const failedItem = dir.findIndex((item) => (typeof item !== 'string'));
    isDir = (failedItem < 0); // valid if failed item not found
  }
  return isDir;
}

/**
 * only minimal implementation of fs.Stats: isDirectory() and isFile()
 * @param path
 */
function statSync(path) {
  const exists = existsSync(path);
  const isDir = (exists && isValidDirectory(path));
  return new Stats(path, exists, isDir);
}

/**
 * @description convertes linux style separators to OS specific separators
 * @param {string} filePath
 * @return {string} converted path
 */
function __correctSeparatorsFromLinux(filePath) {
  const result = filePath.split('/').join(path.sep);
  return result;
}

/**
 * @description - copies list of files from local file system into mock File system
 * @param {array} copyFiles - array of paths (in linux format) relative to source path
 * @param {string} sourceFolder - source folder of files to copy (in linux format)
 * @param {string} mockDestinationFolder - destination folder for copied files {string} in mock File system
 */
function __loadFilesIntoMockFs(copyFiles, sourceFolder, mockDestinationFolder) {
  const mockDestinationFolder_ =  __correctSeparatorsFromLinux(mockDestinationFolder);
  const sourceFolder_ = __correctSeparatorsFromLinux(sourceFolder );
  for (let copyFile of copyFiles) {
    const filePath2 = path.join(sourceFolder_, __correctSeparatorsFromLinux(copyFile));
    let fileData = null;
    const isDir = fsActual.statSync(filePath2).isDirectory();
    if (!isDir) {
      fileData = fsActual.readFileSync(filePath2).toString();
    }
    let dirPath = mockDestinationFolder_;
    fs.ensureDirSync(dirPath);
    const parts = copyFile.split('/');
    const endCount = parts.length - 1;
    for (let i = 0; i < endCount; i++) {
      const part = parts[i];
      dirPath = path.join(dirPath, part);
      fs.ensureDirSync(dirPath);
    }
    if (!isDir) {
      const filePath = path.join(mockDestinationFolder_, parts.join(path.sep));
      // console.log("Copying File: " + filePath);
      fs.writeFileSync(filePath, fileData);
    } else {
      __loadDirIntoMockFs(filePath2, path.join(mockDestinationFolder, copyFile));
    }
  }
}

/**
 * @description - recursively copies folder from local file system into mock File system
 * @param {string} sourceFolder - source folder of files to copy (in linux format)
 * @param {string} mockDestinationFolder - destination folder for copied files {string} in mock File system
 */
function __loadDirIntoMockFs(sourceFolder, mockDestinationFolder) {
  const mockDestinationFolder_ =  __correctSeparatorsFromLinux(mockDestinationFolder);
  fs.ensureDirSync(mockDestinationFolder_);
  const sourceFolder_ = __correctSeparatorsFromLinux(sourceFolder);
  // console.log("Copying Directory: " + sourceFolder_);
  const files = fsActual.readdirSync(sourceFolder_);
  for (let file of files) {
    const sourceFilePath = path.join(sourceFolder, file);
    const mockFilePath = path.join(mockDestinationFolder_, file);
    const isDir = fsActual.statSync(sourceFilePath).isDirectory();
    if (!isDir) {
      const fileData = fsActual.readFileSync(sourceFilePath).toString();
      // console.log("Copying Subfile: " + mockFilePath);
      fs.writeFileSync(mockFilePath, fileData);
    } else {
      // console.log("Entering Subdir: " + mockFilePath);
      __loadDirIntoMockFs( sourceFilePath, mockFilePath);
    }
  }
}

function moveSync(source, destination) {
  copySync(source, destination);
  removeSync(source);
}

fs.__files = () => {
  return mockFS;
};
fs.__dumpMockFS = __dumpMockFS;
fs.__setMockDirectories = __setMockDirectories;
fs.__setMockFS = __setMockFS;
fs.__resetMockFS = __resetMockFS;
fs.__actual = fsActual; // to access actual file system
fs.__loadFilesIntoMockFs = __loadFilesIntoMockFs;
fs.__correctSeparatorsFromLinux = __correctSeparatorsFromLinux;
fs.__loadDirIntoMockFs = __loadDirIntoMockFs;
fs.readdirSync = jest.fn(readdirSync);
fs.readdir = readdir;
fs.writeFileSync = writeFileSync;
fs.readFileSync = jest.fn(readFileSync);
fs.writeJSONSync = outputJsonSync;
fs.outputJsonSync = jest.fn(outputJsonSync);
fs.outputJSONSync = jest.fn(outputJsonSync);
fs.outputJson = outputJson;
fs.outputJSON = outputJson;
fs.readJsonSync = jest.fn(readJsonSync);
fs.readJSONSync = readJsonSync;
fs.readJson = readJson;
fs.readJSON = readJson;
fs.existsSync = jest.fn(existsSync);
fs.exists = exists;
fs.pathExists = exists;
fs.pathExistsSync = jest.fn(existsSync);
fs.outputFileSync = outputFileSync;
fs.removeSync = removeSync;
fs.remove = remove;
fs.copySync = jest.fn(copySync);
fs.renameSync = renameSync;
fs.rename = rename;
fs.ensureDirSync = ensureDirSync;
fs.statSync = statSync;
fs.fstatSync = statSync;
fs.lstatSync = jest.fn(statSync);
fs.moveSync = moveSync;

module.exports = fs;
