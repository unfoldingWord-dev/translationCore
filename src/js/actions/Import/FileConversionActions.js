/**
 * @Description:
 * Actions that call helpers to handle business logic for conversions
**/

export const convert = () => {
  // return((dispatch, getState) => {
  //   FileConversionHelpers.convertToProject(param);
  // });
};
