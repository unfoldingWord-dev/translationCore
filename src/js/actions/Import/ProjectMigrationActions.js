/**
 * @Description:
 * Actions that call helpers to handle business logic for migtaions
**/

export const migrate = () => {
  return((dispatch, getState) => {
  });
};
