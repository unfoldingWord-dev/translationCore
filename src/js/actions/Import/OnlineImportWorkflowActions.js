import path from 'path-extra';
import fs from 'fs-extra';
// actions
import consts from '../../actions/ActionTypes';
import { deleteImportsFolder, deleteProjectFromImportsFolder } from '../../helpers/Import/ProjectImportFilesystemHelpers';
import * as AlertModalActions from '../../actions/AlertModalActions';
import * as OnlineModeConfirmActions from '../../actions/OnlineModeConfirmActions';
import * as ProjectImportStepperActions from '../ProjectImportStepperActions';
import * as MyProjectsActions from '../MyProjects/MyProjectsActions';
import * as ProjectDetailsActions from '../ProjectDetailsActions';
import * as ProjectInformationCheckActions from '../ProjectInformationCheckActions';
// helpers
import * as TargetLanguageHelpers from '../../helpers/TargetLanguageHelpers';
import {
  generateImportPath,
  verifyThisIsTCoreOrTStudioProject,
} from '../../helpers/Import/OnlineImportWorkflowHelpers';
import * as CopyrightCheckHelpers from '../../helpers/CopyrightCheckHelpers';
import {
  getProjectManifest,
  getProjectSaveLocation,
  getTranslate,
  getUsername,
} from '../../selectors';
import * as FileConversionHelpers from '../../helpers/FileConversionHelpers';
import * as ProjectDetailsHelpers from '../../helpers/ProjectDetailsHelpers';
import migrateProject from '../../helpers/ProjectMigration';
import Repo from '../../helpers/Repo';
import { isProjectSupported } from '../../helpers/ProjectValidation/ProjectStructureValidationHelpers';
import {
  openProject,
  closeProject,
  showInvalidVersionError,
} from '../MyProjects/ProjectLoadingActions';
import { delay } from '../../common/utils';
//constants
import { tc_MIN_VERSION_ERROR, IMPORTS_PATH } from '../../common/constants';
import * as ProjectImportFilesystemActions from './ProjectImportFilesystemActions';
import * as ProjectValidationActions from './ProjectValidationActions';

/**
 * Action that dispatches other actions to wrap up online importing
 */
export const onlineImport = () => (dispatch, getState) => new Promise((resolve, reject) => {
  const translate = getTranslate(getState());

  dispatch(OnlineModeConfirmActions.confirmOnlineAction(async () => {
    let importProjectPath = '';
    let link = '';

    try {
      await deleteImportsFolder();
      // Must allow online action before starting actions that access the internet
      link = getState().importOnlineReducer.importLink.trim();
      console.log('onlineImport() - link=' + link);

      dispatch(clearLink());
      // or at least we could pass in the locale key here.
      dispatch(AlertModalActions.openAlertDialog(translate('projects.importing_project_alert', { project_url: link }), true));

      const importPath = await generateImportPath(link);
      console.log('onlineImport() - import to: ' + importPath);

      await fs.ensureDir(importPath);
      console.log('onlineImport() - cloning repo into file system');
      await Repo.clone(link, importPath);
      const selectedProjectFilename = Repo.parseRemoteUrl(Repo.sanitizeRemoteUrl(link)).name;
      console.log('onlineImport() - selectedProjectFilename= ' + selectedProjectFilename);

      dispatch({ type: consts.UPDATE_SELECTED_PROJECT_FILENAME, selectedProjectFilename });
      importProjectPath = path.join(IMPORTS_PATH, selectedProjectFilename);

      // check if we can import the project
      console.log('onlineImport() - check if we can import the project');
      const isValid = verifyThisIsTCoreOrTStudioProject(importProjectPath);

      if (!isValid) {
        const errorMessage = translate('projects.online_import_error', { project_url: link, toPath: importProjectPath });
        console.warn('This is not a valid tStudio or tCore project we can migrate: ', errorMessage);
        throw errorMessage;
      }

      await isProjectSupported(importProjectPath, translate);
      const initialBibleDataFolderName = ProjectDetailsHelpers.getInitialBibleDataFolderName(selectedProjectFilename, importProjectPath);
      await migrateProject(importProjectPath, link, getUsername(getState()));
      // assign CC BY-SA license to projects imported from door43
      await CopyrightCheckHelpers.assignLicenseToOnlineImportedProject(importProjectPath);
      console.log('onlineImport() - start project validation');
      dispatch(ProjectValidationActions.initializeReducersForProjectImportValidation(false));
      await dispatch(ProjectValidationActions.validateProject(importProjectPath));
      const manifest = getProjectManifest(getState());
      const updatedImportPath = getProjectSaveLocation(getState());
      ProjectDetailsHelpers.fixBibleDataFolderName(manifest, initialBibleDataFolderName, updatedImportPath);

      if (!TargetLanguageHelpers.targetBibleExists(updatedImportPath, manifest)) {
        dispatch(AlertModalActions.openAlertDialog(translate('projects.loading_ellipsis'), true));
        console.log('onlineImport() - generate target bible');
        TargetLanguageHelpers.generateTargetBibleFromTstudioProjectPath(updatedImportPath, manifest);
        dispatch(ProjectInformationCheckActions.setSkipProjectNameCheckInProjectInformationCheckReducer(true));
        await delay(200);
        dispatch(AlertModalActions.closeAlertDialog());
        console.log('onlineImport() - validate project');
        await dispatch(ProjectValidationActions.validateProject(updatedImportPath));
      }

      const renamingResults = {};
      await dispatch(ProjectDetailsActions.updateProjectNameIfNecessary(renamingResults));
      const { projectDetailsReducer: { projectSaveLocation } } = getState();

      if (renamingResults.repoRenamed) {
        dispatch({ type: consts.UPDATE_SOURCE_PROJECT_PATH, sourceProjectPath: projectSaveLocation });
        dispatch({ type: consts.UPDATE_SELECTED_PROJECT_FILENAME, selectedProjectFilename: renamingResults.newRepoName });
        await delay(200);
      }
      await dispatch(ProjectImportFilesystemActions.move());

      if (renamingResults.repoRenamed) {
        await dispatch(ProjectDetailsActions.doRenamePrompting());
      }
      dispatch(MyProjectsActions.getMyProjects());

      // TODO: refactor this onlineImport method to remove project opening logic so we are not duplicating logic.

      const finalProjectPath = getProjectSaveLocation(getState());
      console.log('onlineImport() - project import complete: ' + finalProjectPath);
      await dispatch(openProject(path.basename(finalProjectPath), true));
      dispatch(AlertModalActions.closeAlertDialog());
      resolve();
    } catch (error) { // Catch all errors in nested functions above
      console.log('onlineImport() - import error:', error);
      const errorMessage = FileConversionHelpers.getSafeErrorMessage(error, translate('projects.online_import_error', { project_url: link, toPath: importProjectPath }));
      dispatch(recoverFailedOnlineImport(errorMessage));
      reject(errorMessage);
    }
  }));
});

/**
 * Performs recovery actions to cleanup after a failed online import
 * @param {string} errorMessage - A localized error message to show the user.
 * @returns {Function}
 */
export const recoverFailedOnlineImport = (errorMessage) => (dispatch) => {
  // TRICKY: clear last project first to avoid triggering autos-saving.
  dispatch(closeProject());

  if (errorMessage === tc_MIN_VERSION_ERROR) {
    dispatch(showInvalidVersionError());
  } else {
    dispatch(AlertModalActions.openAlertDialog(errorMessage));
  }
  dispatch(ProjectImportStepperActions.cancelProjectValidationStepper());
  dispatch({ type: 'LOADED_ONLINE_FAILED' });
  dispatch(deleteImportProjectForLink());
};

/**
 * TODO: this does not need to be an action.
 * @description - delete project (for link) from import folder
 */
export function deleteImportProjectForLink() {
  return ((dispatch, getState) => {
    const link = getState().importOnlineReducer.importLink;

    if (link) {
      const gitUrl = Repo.sanitizeRemoteUrl(link);
      let project = Repo.parseRemoteUrl(gitUrl);

      if (project) {
        deleteProjectFromImportsFolder(project.name);
      }
    }
  });
}

export function clearLink() {
  return {
    type: consts.IMPORT_LINK,
    importLink: '',
  };
}

export function getLink(importLink) {
  return {
    type: consts.IMPORT_LINK,
    importLink,
  };
}
