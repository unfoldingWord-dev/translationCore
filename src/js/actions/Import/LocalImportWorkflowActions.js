import React from 'react';
import path from 'path-extra';
import ospath from 'ospath';
import fs from 'fs-extra';
import {ipcRenderer} from 'electron';
import consts from '../ActionTypes';
// actions
import * as BodyUIActions from '../BodyUIActions';
import * as AlertModalActions from '../AlertModalActions';
import * as ProjectMigrationActions from '../Import/ProjectMigrationActions';
import * as ProjectValidationActions from '../Import/ProjectValidationActions';
import * as ProjectImportFilesystemActions from './ProjectImportFilesystemActions';
import * as ProjectImportStepperActions from '../ProjectImportStepperActions';
import * as MyProjectsActions from '../MyProjects/MyProjectsActions';
import * as ProjectLoadingActions from '../MyProjects/ProjectLoadingActions';
// helpers
import * as TargetLanguageHelpers from '../../helpers/TargetLanguageHelpers';
import * as FileConversionHelpers from '../../helpers/FileConversionHelpers';
import {getTranslate, getProjectManifest, getProjectSaveLocation} from '../../selectors';
import * as ProjectMergeActions from '../ProjectMergeActions';

// constants
export const ALERT_MESSAGE = (
  <div>
    No file was selected. Please click on the
    <span style={{color: 'var(--accent-color-dark)', fontWeight: "bold"}}>
      &nbsp;Import Local Project&nbsp;
    </span>
    button again and select the project you want to load.
  </div>
);
const IMPORTS_PATH = path.join(ospath.home(), 'translationCore', 'imports');
const PROJECTS_PATH = path.join(ospath.home(), 'translationCore', 'projects');

/**
 * @description Action that dispatches other actions to wrap up local importing
 */
export const localImport = () => {
  return async (dispatch, getState) => {
    const translate = getTranslate(getState());
    // selectedProjectFilename and sourceProjectPath are populated by selectProjectMoveToImports()
    const {
      selectedProjectFilename,
      sourceProjectPath
    } = getState().localImportReducer;
    let importProjectPath = path.join(IMPORTS_PATH, selectedProjectFilename);
    try {
      // convert file to tC acceptable project format
      await FileConversionHelpers.convert(sourceProjectPath, selectedProjectFilename);
      ProjectMigrationActions.migrate(importProjectPath);
      await dispatch(ProjectValidationActions.validate(importProjectPath));
      const manifest = getProjectManifest(getState());
      importProjectPath = getProjectSaveLocation(getState());
      if (!TargetLanguageHelpers.targetBibleExists(importProjectPath, manifest)) {
        TargetLanguageHelpers.generateTargetBibleFromTstudioProjectPath(importProjectPath, manifest);
        await delay(400);
        await dispatch(ProjectValidationActions.validate(importProjectPath));
      }
      const projectName = getState().localImportReducer.selectedProjectFilename;
      const projectPath = path.join(PROJECTS_PATH, projectName);
      if (fs.pathExistsSync(projectPath)) {
        // The project already exists so we handle merging the import data with existing project data,
        // then delete the project dir and then continue the import that copies the import dir to the projects dir
        dispatch(ProjectMergeActions.handleProjectMerge(projectPath, importProjectPath))
        .then(() => {
          dispatch(continueImport());
        })
        .catch(error => {
          dispatch(cancelImport(error));
        });
      } else {
        dispatch(continueImport());
      }
    } catch (error) { // Catch all errors in nested functions above
      const errorMessage = FileConversionHelpers.getSafeErrorMessage(error, translate('projects.import_error', {fromPath: sourceProjectPath, toPath: importProjectPath}));
      dispatch(cancelImport(errorMessage));
    }
  };
};

/**
 * @description selects a project from the filesystem and moves it to tC imports folder.
 * @param startLocalImport - optional parameter to specify new startLocalImport function (useful for testing).
 * Default is localImport()
 */
export const selectLocalProject = (startLocalImport = localImport) => {
  return (dispatch, getState) => {
    return new Promise(async (resolve) => {
      const translate = getTranslate(getState());
      dispatch(BodyUIActions.dimScreen(true));
      dispatch(BodyUIActions.closeProjectsFAB());
      // TODO: the filter name and dialog text should not be set here.
      // we should instead send generic data and load the text in the react component with localization
      // or at least we could insert the locale keys here.
      await delay(500);
      const options = {
        properties: ['openFile'],
        filters: [
          {name: translate('supported_file_types'), extensions: ['usfm', 'sfm', 'txt', 'tstudio', 'tcore']}
        ]
      };
      let filePaths = ipcRenderer.sendSync('load-local', {options: options});
      dispatch(BodyUIActions.dimScreen(false));
      // if import was cancel then show alert indicating that it was cancel
      if (filePaths && filePaths[0]) {
        dispatch(AlertModalActions.openAlertDialog(translate('projects.importing_local_alert'), true));
        const sourceProjectPath = filePaths[0];
        const selectedProjectFilename = path.parse(sourceProjectPath).base.split('.')[0] || '';
        await delay(100);
        dispatch({type: consts.UPDATE_SOURCE_PROJECT_PATH, sourceProjectPath});
        dispatch({type: consts.UPDATE_SELECTED_PROJECT_FILENAME, selectedProjectFilename});
        await dispatch(startLocalImport());
        resolve();
      } else {
        dispatch(AlertModalActions.closeAlertDialog());
        resolve();
      }
    });
  };
};

const continueImport = () => {
  return dispatch => {
    dispatch(ProjectImportFilesystemActions.move())
    .then(() => {
      dispatch(MyProjectsActions.getMyProjects());
      dispatch(ProjectLoadingActions.displayTools());
    })
    .catch(error => {
      dispatch(cancelImport(error));
    });
  };
};

const cancelImport = errorMessage => {
  return dispatch => {
    if (errorMessage) {
      dispatch(AlertModalActions.openAlertDialog(errorMessage));
    }
    // clear last project must be called before any other action.
    // to avoid triggering auto-saving.
    dispatch(ProjectLoadingActions.clearLastProject());
    dispatch(ProjectImportStepperActions.cancelProjectValidationStepper());
    // remove failed project import
    dispatch(ProjectImportFilesystemActions.deleteProjectFromImportsFolder());
  };
};

const delay = (ms) => {
  return new Promise((resolve) =>
    setTimeout(resolve, ms)
  );
};
