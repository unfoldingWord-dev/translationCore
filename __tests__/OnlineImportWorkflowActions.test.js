import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as OnlineImportWorkflowActions from '../src/js/actions/Import/OnlineImportWorkflowActions';
import * as fs from 'fs-extra';
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const importProjectName = 'es-419_tit_text_ulb';
const STANDARD_PROJECT = 'https://git.door43.org/royalsix/' + importProjectName + '.git';

//mocking functions that are relevant to OnlineImportWorkflowActions but not required
jest.mock('../src/js/actions/Import/ProjectMigrationActions', () => ({ migrate: jest.fn() }));
jest.mock('../src/js/actions/Import/ProjectValidationActions', () => ({ validate: () => ({ type: 'VALIDATE' }) }));
jest.mock('../src/js/actions/Import/ProjectImportFilesystemActions', () => ({
  deleteProjectFromImportsFolder: () => ({ type: 'DELETE_PROJECT_FROM_IMORTS' }),
  move: () => ({ type: 'MOVE' })
}));
jest.mock('../src/js/actions/MyProjects/MyProjectsActions', () => ({ getMyProjects: () => ({ type: 'GET_MY_PROJECTS' }) }));
jest.mock('../src/js/actions/MyProjects/ProjectLoadingActions', () => ({
  clearLastProject: () => ({ type: 'CLEAR_LAST_PROJECT' }),
  displayTools: jest.fn(() => ({ type: 'DISPLAY_TOOLS' }))
    .mockImplementationOnce(() => ({ type: 'DISPLAY_TOOLS' }))
    .mockImplementationOnce(() => () => Promise.reject('Some error'))
}));
jest.mock('../src/js/actions/TargetLanguageActions', ()=> ({
  generateTargetBibleFromTstudioProjectPath: () => {}
}));
jest.mock('../src/js/helpers/ProjectValidation/ProjectStructureValidationHelpers', () => ({
  ensureSupportedVersion: () => {}
}));

describe('OnlineImportWorkflowActions.onlineImport', () => {
  let initialState = {};

  beforeEach(() => {
    fs.__resetMockFS();
    initialState = {
      importOnlineReducer: {
        importLink: STANDARD_PROJECT
      },
      settingsReducer: {
        onlineMode: true
      },
      projectDetailsReducer: {
        manifest: {},
        projectSaveLocation: 'project/path'
      }
    };
  });

  it('should import a project that has whitespace in string', () => {
    const expectedActions = [
      { type: 'IMPORT_LINK', importLink: '' },
      {
        type: 'OPEN_ALERT_DIALOG',
        alertMessage: 'projects.importing_project_alert',
        loading: true
      },
      {
        type: 'UPDATE_SELECTED_PROJECT_FILENAME',
        selectedProjectFilename: 'es-419_tit_text_ulb'
      },
      { type: 'VALIDATE' },
      { type: 'MOVE' },
      { type: 'GET_MY_PROJECTS' },
      { type: 'DISPLAY_TOOLS' }
    ];
    const store = mockStore(initialState);
    return store.dispatch(OnlineImportWorkflowActions.onlineImport()).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('on import errors should call required actions', () => {
    const expectedActions = [
      { "importLink": "", "type": "IMPORT_LINK" },
      { "alertMessage": "projects.importing_project_alert", "loading": true, "type": "OPEN_ALERT_DIALOG" },
      { "selectedProjectFilename": "es-419_tit_text_ulb", "type": "UPDATE_SELECTED_PROJECT_FILENAME" },
      { "type": "VALIDATE" }, { "type": "MOVE" }, { "type": "GET_MY_PROJECTS" },
      { "type": "CLEAR_LAST_PROJECT" },
      { "alertMessage": "Some error", "loading": undefined, "type": "OPEN_ALERT_DIALOG" },
      { "showProjectValidationStepper": false, "type": "TOGGLE_PROJECT_VALIDATION_STEPPER" },
      { "type": "CLEAR_LAST_PROJECT" },
      { "type": "CLEAR_COPYRIGHT_CHECK_REDUCER" },
      { "type": "CLEAR_PROJECT_INFORMATION_REDUCER" },
      { "type": "CLEAR_MERGE_CONFLICTS_REDUCER" },
      { "type": "RESET_PROJECT_VALIDATION_REDUCER" },
      { "type": "GET_MY_PROJECTS" },
      { "type": "DELETE_PROJECT_FROM_IMORTS" },
      { "type": "LOADED_ONLINE_FAILED" },
      { "type": "DELETE_PROJECT_FROM_IMORTS" }
    ];
    const store = mockStore(initialState);
    return store.dispatch(OnlineImportWorkflowActions.onlineImport()).catch((error) => {
      expect(error).toEqual('Some error');
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});
