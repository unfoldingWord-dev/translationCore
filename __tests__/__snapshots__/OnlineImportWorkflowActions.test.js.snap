// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnlineImportWorkflowActions.onlineImport on import errors should call required actions 1`] = `
Array [
  Object {
    "importLink": "",
    "type": "IMPORT_LINK",
  },
  Object {
    "alertMessage": "projects.importing_project_alert",
    "loading": true,
    "type": "OPEN_ALERT_DIALOG",
  },
  Object {
    "selectedProjectFilename": "es-419_tit_text_ulb",
    "type": "UPDATE_SELECTED_PROJECT_FILENAME",
  },
  Object {
    "type": "RESET_PROJECT_VALIDATION_REDUCER",
  },
  Object {
    "type": "CLEAR_PROJECT_INFORMATION_REDUCER",
  },
  Object {
    "nextDisabled": true,
    "type": "UPDATE_PROJECT_VALIDATION_NEXT_BUTTON_STATUS",
  },
  Object {
    "type": "VALIDATE",
  },
  Object {
    "skipProjectNameCheck": true,
    "type": "SET_SKIP_PROJECT_NAME_CHECK_IN_PROJECT_INFORMATION_CHECK_REDUCER",
  },
  Object {
    "type": "VALIDATE",
  },
  Object {
    "pathLocation": "project/__book",
    "type": "SET_SAVE_PATH_LOCATION",
  },
  Object {
    "sourceProjectPath": "project/path",
    "type": "UPDATE_SOURCE_PROJECT_PATH",
  },
  Object {
    "selectedProjectFilename": "__book",
    "type": "UPDATE_SELECTED_PROJECT_FILENAME",
  },
  Object {
    "type": "MOVE",
  },
  Object {
    "alertMessage": "projects.renamed_project",
    "button1Text": "buttons.ok_button",
    "button2Text": undefined,
    "buttonLinkText": null,
    "callback": [Function],
    "type": "OPEN_OPTION_DIALOG",
  },
  Object {
    "type": "GET_MY_PROJECTS",
  },
  Object {
    "type": "CLEAR_LAST_PROJECT",
  },
  Object {
    "alertMessage": "Some error",
    "loading": undefined,
    "type": "OPEN_ALERT_DIALOG",
  },
  Object {
    "showProjectValidationStepper": false,
    "type": "TOGGLE_PROJECT_VALIDATION_STEPPER",
  },
  Object {
    "type": "CLEAR_LAST_PROJECT",
  },
  Object {
    "type": "CLEAR_COPYRIGHT_CHECK_REDUCER",
  },
  Object {
    "type": "CLEAR_PROJECT_INFORMATION_REDUCER",
  },
  Object {
    "type": "CLEAR_MERGE_CONFLICTS_REDUCER",
  },
  Object {
    "type": "RESET_PROJECT_VALIDATION_REDUCER",
  },
  Object {
    "type": "GET_MY_PROJECTS",
  },
  Object {
    "type": "DELETE_PROJECT_FROM_IMORTS",
  },
  Object {
    "type": "LOADED_ONLINE_FAILED",
  },
  Object {
    "type": "DELETE_PROJECT_FROM_IMORTS",
  },
]
`;

exports[`OnlineImportWorkflowActions.onlineImport should import a project that has whitespace in string 1`] = `
Array [
  Object {
    "importLink": "",
    "type": "IMPORT_LINK",
  },
  Object {
    "alertMessage": "projects.importing_project_alert",
    "loading": true,
    "type": "OPEN_ALERT_DIALOG",
  },
  Object {
    "selectedProjectFilename": "es-419_tit_text_ulb",
    "type": "UPDATE_SELECTED_PROJECT_FILENAME",
  },
  Object {
    "type": "RESET_PROJECT_VALIDATION_REDUCER",
  },
  Object {
    "type": "CLEAR_PROJECT_INFORMATION_REDUCER",
  },
  Object {
    "nextDisabled": true,
    "type": "UPDATE_PROJECT_VALIDATION_NEXT_BUTTON_STATUS",
  },
  Object {
    "type": "VALIDATE",
  },
  Object {
    "skipProjectNameCheck": true,
    "type": "SET_SKIP_PROJECT_NAME_CHECK_IN_PROJECT_INFORMATION_CHECK_REDUCER",
  },
  Object {
    "type": "VALIDATE",
  },
  Object {
    "pathLocation": "project/__book",
    "type": "SET_SAVE_PATH_LOCATION",
  },
  Object {
    "sourceProjectPath": "project/path",
    "type": "UPDATE_SOURCE_PROJECT_PATH",
  },
  Object {
    "selectedProjectFilename": "__book",
    "type": "UPDATE_SELECTED_PROJECT_FILENAME",
  },
  Object {
    "type": "MOVE",
  },
  Object {
    "alertMessage": "projects.renamed_project",
    "button1Text": "buttons.ok_button",
    "button2Text": undefined,
    "buttonLinkText": null,
    "callback": [Function],
    "type": "OPEN_OPTION_DIALOG",
  },
  Object {
    "type": "GET_MY_PROJECTS",
  },
  Object {
    "type": "DISPLAY_TOOLS",
  },
]
`;
